app/src/main/java/com/aletheiaware/perspective/android/WorldAdapter.java

```diff
@@ -102,31 +109,77 @@ public void onClick(View v) {
-        if (names.isEmpty()) {
-            holder.setEmptyView();
+    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
+        if (position == 0) {
+            StarCountHolder sch = (StarCountHolder) holder;
+            int earned = 0;
+            int max = 0;
+            for (String name : names) {
+                Integer i = starsMap.get(name);
+                if (i != null) {
+                    earned += i;
+                }
+                World w = worldsMap.get(name);
+                if (w != null) {
+                    max += w.getPuzzleCount() * PerspectiveAndroidUtils.MAX_STARS;
+                }
+            }
+            sch.setStars(earned, max);
         } else {
-            String name = names.get(position);
-            World world = worldsMap.get(name);
-            if (world == null) {
-                holder.set(name, pricesMap.get(name));
+            WorldViewHolder wvh = (WorldViewHolder) holder;
+            if (names.isEmpty()) {
+                wvh.setEmptyView();
             } else {
-                int s = 0;
-                Integer stars = starsMap.get(name);
-                if (stars != null) {
-                    s = stars;
+                String name = names.get(position - 1);
+                World world = worldsMap.get(name);
+                if (world == null) {
+                    wvh.set(name, pricesMap.get(name));
+                } else {
+                    int s = 0;
+                    Integer stars = starsMap.get(name);
+                    if (stars != null) {
+                        s = stars;
+                    }
+                    int puzzles = world.getPuzzleCount();
+                    int worldStars = 0;
+                    if (puzzles > 0) {
+                        worldStars = s / puzzles;
+                    }
+                    wvh.set(world, worldStars);
                 }
-                holder.set(world, s);
             }
         }
     }
 
+    @Override
+    public int getItemViewType(int position) {
+        if (position == 0) {
+            return TYPE_HEADER;
+        }
+        return TYPE_ITEM;
+    }
+
     @Override
     public int getItemCount() {
+        int count = 1; // Header
         if (names.isEmpty()) {
-            return 1;// For empty view
+            return count + 1;// Empty view
+        }
+        return count + names.size();
+    }
+
+    static class StarCountHolder extends RecyclerView.ViewHolder {
+
+        private final TextView starCount;
+
+        StarCountHolder(View view) {
+            super(view);
+            starCount = view.findViewById(R.id.world_star_count);
+        }
+
+        void setStars(int earned, int max) {
+            starCount.setText(starCount.getContext().getString(R.string.star_count_format, earned, max));
         }
-        return names.size();
     }
 
     static class WorldViewHolder extends RecyclerView.ViewHolder {
```

Comment:
Do you mean 'itemStar1', 'itemStar2'... could all be in one array 'itemStars'? It could be, I'm not sure whether that would be any more readable;

```
private final View[] itemStars = new View[PerspectiveAndroidUtils.MAX_STARS];

#init
itemStars[0] = findViewById(R.id.world_item_star1);
itemStars[1] = findViewById(R.id.world_item_star2);
itemStars[2] = findViewById(R.id.world_item_star3);
itemStars[3] = findViewById(R.id.world_item_star4);
itemStars[4] = findViewById(R.id.world_item_star5);

#set
for (int i = 0; i < PerspectiveAndroidUtils.MAX_STARS; i++) {
    itemStar[i].setVisibility(stars > i ? View.VISIBLE : View.INVISIBLE);
}

#set
for (int i = 0; i < PerspectiveAndroidUtils.MAX_STARS; i++) {
    itemStar[i].setVisibility(View.GONE);
}

#setEmptyView
for (int i = 0; i < PerspectiveAndroidUtils.MAX_STARS; i++) {
    itemStar[i].setVisibility(View.GONE);
}
```

The compiler would unroll the loops anyway so I don't think there is a performance benefit.

Thoughts?

Reply:
Yes, I think this is "better" less repetition of itemStars_X and the last two set loops can be simpler:
```
for (item : itemStars) {
  item.setVisibility(View.Gone);
}
```
I suspect there is a way to initialize itemStars with a for-each loop too, but that pesky findViewById(R.id.world_item_starX) needs to be "iterable".

Anyway, not a big deal just seemed like a lot of similar code.

Reply:
Okay, done. The findViewById could be iterable by declaring a resource array, but that seems like overkill.